### GROUP 6 TEAM TASK

Build a regression model to predict wedding costs from all the available variables.

Note the "Payor" column contains a categorical variable, which takes two values – "Bride's family" and "Groom's family". To use the variable in the regression model you need to first convert it to a numerical variable with 0 representing one category and 1 representing the other (use the replace function in Excel). Analyse residuals to confirm that the assumptions of linear regression are valid for this model.

Interpret and comment on the regression results:
* what is the quality of the model?
* which predictor variables are the most important?


In [14]:
# IMPORT REQUIRED LIBRARIES
import pandas as pd
import statsmodels.api as sm

# USE BOKEH FOR THE VISUALIZATIONS
from bokeh.io import output_notebook
output_notebook()
from bokeh.plotting import figure
from bokeh.io import show

# READ WEDDINGS XLSX INTO A PANDAS DATAFRAME
df = pd.read_excel("Weddings.xlsx")

# CREATE A FUNCTION THAT ITERATES THROUGH THE 
# PAYOR COLUMN AND CHANGES BRIDE TO A 0
# AND THE GROOM TO A 1
def func(row):
    if row['Payor'] == r"Bride's Parents":
        row['Payor'] = 0
    elif row['Payor'] == r"Groom's Parents":
        row['Payor'] = 1
    return row

# APPLY THE FUNCITON
df = df.apply(func, axis=1)
# SHOW THE DATAFRAME HEAD
df.head(3)

Loading BokehJS ...

CoupleIncome  BrideAge  Payor  WeddingCost  Attendance
0        130000        22      0        60700         300
1        157000        23      0        52000         350
2         98000        27      1        47000         150

In [15]:
# CREATE A MULTIPLE REGRESSION MODEL BASED ON ATTENDANCE, BRIDE AGE, INCOME, AND PAYOR
model = sm.OLS.from_formula(
    'WeddingCost ~ Attendance + \
                    BrideAge + \
                    CoupleIncome + \
                    Payor', data=df).fit()

In [16]:
# PRINT  THE SUMMARY OF THE MODEL
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:            WeddingCost   R-squared:                       0.754
Model:                            OLS   Adj. R-squared:                  0.705
Method:                 Least Squares   F-statistic:                     15.31
Date:                Sat, 13 Feb 2021   Prob (F-statistic):           6.97e-06
Time:                        11:23:12   Log-Likelihood:                -254.71
No. Observations:                  25   AIC:                             519.4
Df Residuals:                      20   BIC:                             525.5
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept    -3048.3092   1.16e+04     -0.263   

### (1) Coefficients on the variables.

wedding_cost = -3048.3092 + 54.1315 ∗ Attendance -233.311 ∗ BrideAge + 0.3238 ∗ CoupleIncome + 3712.87 ∗ Payor.

**If a couple is planning a wedding for 175 guests, the couple's income is $100,000, the bride's age is 33, and the payor is the Bride's family, how much should they budget?**

**wedding_cost = 34 818 USD**

### (2) Significance of the variables. 

Attendance has a significant impact on the wedding cost
BrideAge has no significance on the wedding cost
Couple Income has a significant impact on the wedding cost
Payor has no significant impact on the wedding cost

### (3) Quality of the model.
R Sqaured and Adjusted R Sqaured have an average of 73% accuracy, so the additional variables have helped to improve the model.

## Checking the assumptions of normality and zero mean of residuals

As with the simple regression model, we can plot the standardized residuals and their histogram to confirm that the assumptions of normality of the distribution of residuals and of the zero mean of residuals are valid with this model.

In [17]:
fig = figure(height=400, width=400)
# THE X AXIS IS THE FITTED VALUES
# THE Y AXIS IS THE STANDARDIZED RESIDUALS
st_resids = model.get_influence().resid_studentized_internal
fig.circle(model.fittedvalues, st_resids)

show(fig)

In [18]:
import numpy as np
# CREATE A HISTOGRAM WITH 10 BINS
hist, edges = np.histogram(st_resids, bins=10)

In [19]:
fig = figure(height=400, width=400)
fig.quad(top=hist, bottom=0, left=edges[:-1], right=edges[1:], line_color="white")
show(fig)

The scatterplot and the histogram suggest the residuals are equally distributed around 0 and are normally distributed. The results of the Jarque-Bera test on the residuals (the third table of the summary) also indicate that the errors are distributed normally: the p-value equals 0.12, therefore we fail to reject the null hypothesis of normal distribution.

As with the simple linear regression model, we do not test for independence of the errors and for homoskedasticity, as these assumptions are likely to be violated only if the observations are ordered along a temporal dimension.

Thus, the addition of the extra variables not only improved the quality of the model, but also ensured that the assumptions of the classical linear regression method hold with the model.